docs/source/configuration/validation.md

```diff
@@ -0,0 +1,273 @@
+---
+myst:
+  html_meta:
+    "description": "Client side form field validation"
+    "property=og:description": "Client side form field validation"
+    "property=og:title": "Form fields validation"
+    "keywords": "Volto, Plone, frontend, React, configuration, form, fields, validation"
+---
+
+# Client side form field validation
+
+Volto provides a mechanism for delivering form field validation in an extensible way.
+This extensibility is based on the Volto registry.
+It applies to content types, custom programatically generated forms and blocks schema settings.
+All of them are serialized using JSON schema standard and then Volto generates the resultant form out of it.
+
+## Registering a validator
+
+You can register a validator using the registry API from your add-on configuration.
+The validators are registered using the `registerUtility` API method.
+
+### Registering and declaring a simple validator
+
+The most common thing is to have a field that you want to validate with a specific validator.
+Volto provide some default validators, see at the end of this chapter for more information.
+
+#### For Volto custom forms and block schema forms
+
+When you define them programatically, in your core, using JSON schema, you can register a custom validator using the `format` property.
+This is the case of creating the schema for a block:
+
+```ts
+let blockSchema = {
+  // ... fieldset definition in here
+  properties: {
+    ...schema.properties,
+    customField: {
+      title: 'My custom URL field',
+      description: '',
+      format: 'url'
+    },
+  },
+  required: [],
+};
+```
+
+`url` named validator should be registered by this name as a Volto validator utility:
+
+```ts
+config.registerUtility({
+  type: 'validator',
+  name: 'url',
+  method: urlValidator,
+})
+```
+
+In this case, the `urlValidator` method validator will be applied for the block field `customField`.
+
+#### For content types
+
+Content types also can specify the `format` using the schema hints in the backend in the `frontendOptions`:
+
+```python
+from plone.supermodel import model
+from zope import schema
+
+class IMyContent(model.Schema):
+    directives.widget(
+        "customField",
+        frontendOptions={
+            "format": "url",
+        },
+    )
+    customField = schema.TextLine(
+        title="Custom URL field",
+        required=False,
+    )
+    # Rest of your content type definition
+```
+
+For the record, the resultant `plone.restapi` response will be something like the following, a bit different than in blocks JSON schema.
+But the validation engine will take care too:
+
+```json
+{
+  "properties": {
+    "customField": {
+      "title": "Custom URL field",
+      "widgetOptions": {
+        "frontendOptions": {
+          "format": "url"
```

Comment:
Can we change the api call so it will be "flattened" to the root of the field so we only have 1 way specifying the format?

Reply:
I had this wish back in the day too, but it never happened, since we already had this "scape hatch" for all customizable things in Python schema hints. Maybe we can push again in the backend side for this to happen.

@davisagli do you think we could try to unify this also in the backend? Or do you remember the reason behind we never did it?

Reply:
It depends on the property. For things that are valid JSON Schema, it would make sense to have plone.restapi serialize them at the top level. For other properties that are not part of JSON Schema, it makes sense to keep them inside the widgetOptions.